On chain DCAP verification

Niederb opened the pull request.

clangenb commented on primitives/teerex/Cargo.toml:
```diff
@@ -9,26 +9,24 @@ edition = "2021"
 
 [dependencies]
 codec = { version = "3.0.0", default-features = false, features = ["derive"], package = "parity-scale-codec" }
+common-primitives = { path = "../common", default-features = false }
```
We miss the `common-primitives/std` here.

clangenb commented on primitives/teerex/src/lib.rs:
```diff
@@ -45,10 +56,139 @@ impl<PubKey, Url> Enclave<PubKey, Url> {
+		Self { issue_date, next_update, tcb_levels }
+	}
+
+	pub fn is_valid(&self, examinee: &TcbVersionStatus) -> Option<TcbVersionStatus> {
```
Above you have defined the functino slightly different: `fn is_valid() -> bool` and here it returns an option. Is there any reason for that?

OverOrion replied:
I think it might be that the onchain _database_ may not contain the `examinee`, is that so @Niederb?

Niederb replied:
Actually the `Option` was really not needed as it was only checked that it is `Some`. I changed the function to return `bool`. I also renamed both functions to `verify_examinee(..)` to emphasize the fact that it is the parameter that is being verified (and not `Self`). I hope this makes it a bit clearer.

clangenb replied:
Perfect!

clangenb commented on primitives/teerex/src/lib.rs:
```diff
@@ -45,10 +56,139 @@ impl<PubKey, Url> Enclave<PubKey, Url> {
+
+	#[test]
+	fn tcb_full_is_valid() {
+		let reference = TcbVersionStatus::new(hex!("11110204018007000000000000000000"), 7);
```
What kind of reference is this?

clangenb commented on teerex/sgx-verify/Cargo.toml:
```diff
@@ -1,20 +1,28 @@
 base64 = { version = "0.13", default-features = false, features = ["alloc"] }
-chrono = { version = "0.4", default-features = false }
+chrono = { version = "0.4", default-features = false, features = ["serde"] }
 codec = { version = "3.0.0", default-features = false, features = ["derive"], package = "parity-scale-codec" }
+der = { default-features = false, version = "0.6.0" }
+hex = { default-features = false, version = "0.4.3", features = ["alloc"] }
+ring = { git = "https://github.com/Niederb/ring-xous.git", branch = "0.16.20-cleanup", default-features = false, features = ["alloc"] }
 scale-info = { version = "2.0.1", default-features = false, features = ["derive"] }
+serde = { default-features = false, version = "1.0.140", features = ["derive"] }
 serde_json = { version = "1.0", default-features = false, features = ["alloc"] }
-webpki = { default-features = false, git = "https://github.com/scs/webpki-nostd.git" }
+webpki = { default-features = false, git = "https://github.com/scs/webpki-nostd.git", branch = "tn/use-ring-xous" }
+x509-cert = { default-features = false, version = "0.1.0", features = ["alloc"] }
```
There are some missing std-enablements.

clangenb commented on teerex/sgx-verify/fuzz/Cargo.lock:
```diff
@@ -0,0 +1,3445 @@
+# This file is automatically @generated by Cargo.
```
This cargo lock can be removed if you add the `sgx-verify-fuzz` to the workspace.

Niederb replied:
The setup was automatically created by `cargo-fuzz` and they create a workspace on purpose. I'm not sure how this would affect their CLI-tools. Therefore I think I would stick to their setup. Or do you have a strong opinion?
```
# Prevent this from interfering with workspaces
[workspace]
members = ["."]
```

clangenb replied:
Usually, I have, but I guess they will have their reasons. So we can leave it like that.

clangenb commented on teerex/sgx-verify/fuzz/fuzz_targets/signature_check.rs:
```diff
@@ -0,0 +1,29 @@
+#![no_main]
+#![feature(core_panic)]
+
+pub extern crate alloc;
```
Very happy to see some fuzzing!

Niederb replied:
It is a first step, but to provide real value we would need CI integration. I added a comment regarding what is missing in #99.

clangenb replied:
I agree. Still, thanks for the start!

clangenb reviewed (commented):
I just realized that I am reviewing the wrong PR. So I halfway now and change to the one from the worker. :D

clangenb reviewed (commented):
Looks very good, at few places I was tempted to check suggest some functional recommendations for rust, as it was fun to do. :D

clangenb commented on teerex/sgx-verify/src/collateral.rs:
```diff
@@ -0,0 +1,262 @@
+impl TcbInfo {
+	/// This extracts the necessary information into the struct that we actually store in the chain
+	pub fn to_chain_tcb_info(&self) -> (Fmspc, TcbInfoOnChain) {
+		let mut valid_tcbs: Vec<TcbVersionStatus> = Vec::new();
```
The functional approach tends to be more efficient than the `for .. in` loops.

Niederb replied:
Nice. You are the iterator master 🙇 😸

clangenb replied:
I must admit that I had to verify that this compiles. :D

OverOrion commented on teerex/sgx-verify/src/collateral.rs:
```diff
@@ -203,7 +203,7 @@ pub struct TcbInfo {
 
 impl TcbInfo {
 	/// This extracts the necessary information into the struct that we actually store in the chain
```
I think the doc comment should be updated to reflect the change in the return value.

Niederb replied:
I don't understand what you mean by "change in the return value". You refer to the two methods having the same comment?

OverOrion replied:
Sorry for that, let me clarify it.
Previously this function returned a single value `TcbInfoOnChain` (e.g. `the struct that we actually store in the chain` ), but now it returns the tuple `(Fmspc, TcbInfoOnChain)`.
I understand that the `Fmspc` will be the key and the `TcbInfoOnChain` will be the value, but the comment saying otherwise could be misleading.
How about this?
```suggestion
	/// This extracts the necessary information into a tuple (`(Key, Value)`) that we actually store in the chain.
```

OverOrion commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -407,35 +407,40 @@ pub fn verify_certificate_chain<'a>(
-	let mut dcap_quote_clone = dcap_quote;
-	let q: DcapQuote =
+	qe: &QuotingEnclave,
+) -> Result<([u8; 6], TcbVersionStatus, SgxReport), &'static str> {
```
I think the `Fmspc` type should be used instead of the `[u8; 6]`.

OverOrion commented on teerex/src/tests/test_cases.rs:
```diff
@@ -83,6 +83,32 @@ fn register_quoting_enclave_works() {
+#[test]
+fn register_tcb_info_works() {
+	new_test_ext().execute_with(|| {
+		Timestamp::set_timestamp(1671606747000);
```
Could you please add some comments for this magic number?
Like so:
`// this is 2022-11-17T12:45:32Z`

OverOrion commented on teerex/src/tests/test_cases.rs:
```diff
@@ -83,6 +83,32 @@ fn register_quoting_enclave_works() {
+		));
+		let fmspc = hex!("00906EA10000");
+		let tcb_info = Teerex::tcb_info(fmspc);
+		assert_eq!(tcb_info.next_update, 1681649132000);
```
Could you please add some comments for this magic number?

OverOrion commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -134,6 +134,10 @@ impl Decode for QeCertificationData {
 		let size = u32::from_le_bytes(size_buf);
+		// This is an arbitrary limit to prevent out of memory issues. Intel does not specify a max value
+		if size > 65_000 {
+			return Result::Err(codec::Error::from("Certification data too long"))
```
How about returning the size and limit in the error message? Hopefully nobody would encounter it, but just to be as user (/dev) friendly as possible :)

OverOrion commented on teerex/sgx-verify/src/tests.rs:
```diff
@@ -219,10 +219,21 @@ fn verify_tcb_info_signature() {
 }
 
+/// This is demo code of how a CRL certificate can be parsed and how the revoked serials can be extracted
+/// The part that is missing/open is how to verify the certificate chain of the CRL
```
How about adding a TODO comment here?

OverOrion commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -431,20 +431,24 @@ pub fn verify_dcap_quote(
 	let (fmspc, tcb_info) = extract_tcb_info(&certs[0])?;
 
-	const AUTHENTICATION_DATA_SIZE: usize = 32; // This is actually variable but assume 32 for now
+	// For this part some understanding of the document (Especially chapter A.4: Quote Format)
+	// Intel® Software Guard Extensions (Intel® SGX) Data Center Attestation Primitives: ECDSA Quote Library API
+	// https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
```
Thank you for this! :100:

OverOrion reviewed (commented):
Overall a very nice job :100: , I did the review commit-by-commit, so if a note no longer applies, please comment on it and I will double check!

Niederb commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -629,3 +636,80 @@ pub fn verify_server_cert(
+
+fn get_fmspc(der: &[u8]) -> Result<Vec<u8>, &'static str> {
+	let bytes_oid = FMSPC_OID.as_bytes();
+	let mut offset = der
+		.windows(bytes_oid.len())
+		.position(|window| window == bytes_oid)
+		.ok_or("Certificate does not contain 'FMSPC_OID'")?;
+	offset += 12; // length oid (10) + asn1 tag (1) + asn1 length10 (1)
```
The thing is that there are three offsets that are slightly different already in this file and then two more in the IAS code. Writing a generic one will be tricky and extracting only the part that is shared will not bring us much in terms of abstraction. I don't think this is worth it.

Niederb commented on teerex/ias-verify/src/collateral.rs:
```diff
@@ -111,12 +113,41 @@ pub struct EnclaveIdentity {
-	mrsigner: String,
+	#[serde(deserialize_with = "decode_mrsigner")]
+	#[serde(serialize_with = "encode_mrsigner")]
+	mrsigner: [u8; 32],
```
I added a type alias for mrsigner and mrenclave. For mrsigner the situation is sometimes a bit fuzzy because it is also used as a public key in the blockchain domain (for example to sign an extrinsic).

Niederb commented on teerex/ias-verify/src/lib.rs:
```diff
@@ -53,6 +53,35 @@ pub struct TcbLevel {
+	sgxtcbcomp14svn: u8,
+	sgxtcbcomp15svn: u8,
+	sgxtcbcomp16svn: u8,
+	pcesvn: u8,
```
Yes, that was the reason but that is actually one nice thing they changed with the API version 4: it can now be easily treated as array.

Niederb commented on teerex/src/lib.rs:
```diff
@@ -166,9 +178,92 @@ pub mod pallet {
+			// Quoting enclaves are registered globally and not for a specific sender
+			let _sender = ensure_signed(origin)?;
+			let quoting_enclave =
+				Self::verify_quoting_enclave(enclave_identity, signature, certificate_chain)?;
+			<QuotingEnclaveRegistry<T>>::put(quoting_enclave);
```
`DCAP_SERVER_ROOTS` is basically our equivalent of pre-installed root certificates in a browser. During verification of the certificate chain we then verify that the chain is "connected" to a root certificate we trust. 
The setup is basically the same for IAS except it is a different root certificate (I don't know why exactly). Also see the `IAS_SERVER_ROOTS` just above in the code.

clangenb replied:
So the `DCAP_SERVER_ROOTS` is a well-known intel root-certificate, but it is not something that is related to the infrastructure that hosts the PCCS?

Niederb replied:
Exactly. Unless Intel decides to change the root certificate this should work, but if they change then I assume all their verification enclaves will also start failing.

clangenb replied:
Ok, perfect. Thanks for the explanation!

Niederb commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -0,0 +1,729 @@
+		.iter()
+		.filter(|e| e.extn_id == INTEL_SGX_EXTENSION_OID)
+		.map(|e| e.extn_value.to_vec())
+		.collect();
+	ensure!(ext.len() == 1, "There should only be one section containing Intel extensions");
```
Oh, very nice! But what language is that with `isNone()` and `isSome()`? 😆

clangenb replied:
hahaha, this is where I show that I had an affair with dart :P

clangenb commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -0,0 +1,729 @@
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+#[repr(C)]
+pub struct SgxReportBody {
+	cpu_svn: [u8; 16],    /* (  0) Security Version of the CPU */
+	misc_select: [u8; 4], /* ( 16) Which fields defined in SSA.MISC */
```
Ok, then we can keep it for now, but in general, I prefer doc strings. :)

clangenb commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -0,0 +1,729 @@
+	offset += 1; // length_from_raw_data does not move the offset when the length is encoded in a single byte
+	ensure!(len == 1 || len == 2, "PCESVN must be 1 or 2 bytes");
+	let data = der.get(offset..offset + len).ok_or("Index out of bounds")?;
+	if data.len() == 1 {
+		Ok(u16::from(data[0]))
+	} else {
+		// Unwrap is fine here as we check the length above
+		// DER integers are encoded in big endian
+		Ok(u16::from_be_bytes(data.try_into().unwrap()))
+	}
```
Ahhhh, I see. I did not expect this to return an error. Makes total sense, thanks for explaining!

clangenb commented on teerex/ias-verify/src/lib.rs:
```diff
@@ -35,16 +38,98 @@ mod utils;
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+#[repr(C)]
+pub struct DcapQuoteHeader {
```
I wow, I did not know about mobile coin. They made some pretty impressive rust stuff around the intel SDK. They did something that I always had in mind; abstracting away make from the build process.

Anyhow, regarding the types, we have a dealbreaker, as they do not derive the scale codec, which we need.

Niederb replied:
The crate certainly looks interesting. I would suggest to open a ticket to look into it a bit more. During implementation I thought it would also be nice if we could separate the whole attestation code more from the substrate, such that it works without dependencies to substrate. Then we could maybe manage without the scale codec. Maybe that could be another ticket?

Niederb commented on teerex/ias-verify/src/lib.rs:
```diff
@@ -185,14 +272,138 @@ pub static IAS_SERVER_ROOTS: webpki::TLSServerTrustAnchors = webpki::TLSServerTr
+	let intermediate_slices: Vec<&[u8]> = intermediate_certs.iter().map(Vec::as_slice).collect();
+
+	let sig_algs = &[&webpki::ECDSA_P256_SHA256];
+	println!("Intermediate: {}", intermediate_slices[1..2].len());
```
I don't think I saw any requirements regarding the length of the certificate chain. In this case I already removed the limitation but in the enclave registration I check for length:
`ensure!(3 == certs.len(), "Certificate chain must have 3 certificates");`
I think I will remove this check. What do you think?

OverOrion replied:
I agree with you, please do so :100:

Niederb replied:
I check now that there are at least two certificates. The same as in the collateral registration:
`ensure!(certs.len() >= 2, "Certificate chain must have at least two certificates");`

Niederb commented on teerex/src/lib.rs:
```diff
@@ -392,6 +440,29 @@ impl<T: Config> Pallet<T> {
+			.map_err(|_| <Error<T>>::EnclaveSignerDecodeError)?;
+		ensure!(sender == &enclave_signer, <Error<T>>::SenderIsNotAttestedEnclave);
+
+		// TODO: activate state checks as soon as we've fixed our setup
+		// ensure!((report.status == SgxStatus::Ok) | (report.status == SgxStatus::ConfigurationNeeded),
+		//     "RA status is insufficient");
+		// log::info!("teerex: status is acceptable");
```
Yes. See issue #83

OverOrion replied:
I see, what a bummer :cry:.
How about this?
```suggestion
		// TODO: activate state checks as soon as we've fixed our setup #83
		// ensure!((report.status == SgxStatus::Ok) | (report.status == SgxStatus::ConfigurationNeeded),
		//     "RA status is insufficient");
		// log::info!("teerex: status is acceptable");
```

Niederb commented on teerex/src/lib.rs:
```diff
@@ -166,6 +167,53 @@ pub mod pallet {
+			})?;
+
+			#[cfg(not(feature = "skip-ias-check"))]
+			if !<AllowSGXDebugMode<T>>::get() && enclave.sgx_mode == SgxBuildMode::Debug {
```
Yes, exactly. The same check is done for the IAS registration.

Niederb commented on teerex/sgx-verify/src/lib.rs:
```diff
@@ -0,0 +1,729 @@
+		let mut certification_data = vec![0; size.try_into().unwrap()];
+		input.read(&mut certification_data)?;
+
+		Ok(Self { certification_data_type, size, certification_data })
```
Is there a way to "plug" custom traits int the scale codec? Because this is only a single field in a big hierarchical structure. For the most part this matches with what `Encode`/`Decode` does but unfortunately not in these two cases.

clangenb replied:
Ah, I see. No, this is actually not possible, I guess we leave it as it is then. Thanks for the explanation!

Niederb commented:
@clangenb @OverOrion I think I addressed all the issues. If I did not reply I either thought that the comment was outdated or that I fixed it in code.

clangenb reviewed (approved):
Thanks, this looks good to me now!

OverOrion commented on teerex/sgx-verify/fuzz/fuzz_targets/decode_quote.rs:
```diff
@@ -0,0 +1,11 @@
+fuzz_target!(|data: &[u8]| {
+	let mut copy = data;
+	let quote: Result<DcapQuote, codec::Error> = Decode::decode(&mut copy);
+	//assert!(quote.is_err());
```
I think this should not be commented out.

Niederb replied:
Good point, but the this needs to be commented, because the fuzzer actually manages to find byte patterns that can be deserialized. This is not so difficult as there are basically no checks at this point. 
I added a comment that should clarify this.
The test is to check for panics in `decode()`.

OverOrion reviewed (approved):
Overall LGTM (there is a commented out fuzzing test case), thank you so much for it! :rocket:

OverOrion reviewed (approved):
LGTM, thanks for the note!
